core/src/main/java/com/oath/oak/Slice.java

```diff
@@ -6,48 +6,164 @@
 
 package com.oath.oak;
 
+import sun.nio.ch.DirectBuffer;
+
 import java.nio.ByteBuffer;
 
-// Slice is a "small part" of a bigger block of the underlying managed memory.
-// Slice is allocated for data (key or value) and can be de-allocated later
+// Represents a portion of a bigger block which is part of the underlying managed memory.
+// It is allocated via block memory allocator, and can be de-allocated later
 class Slice {
-    private final int blockID;
-    private final ByteBuffer buffer;
-    // version with which this slice was allocated, if slice is not a result of a new creation the version can be invalid
-    private int version;
-    // This field is only used for sanity checks purposes and it should not be used, nor changed.
-    private final int originalPosition;
-
-    public Slice(int blockID, ByteBuffer buffer) {
+
+    protected int blockID;
+    protected int offset;
+    protected int length;
+
+    // Allocation time version
+    protected int version;
+
+    protected ByteBuffer buffer;
+
+    public Slice() {
+        invalidate();
+    }
+
+    // Used to duplicate the allocation state. Does not duplicate the buffer itself.
+    Slice(Slice otherSlice) {
+        copyFrom(otherSlice);
+    }
+
+    // Reset to invalid state
+    void invalidate() {
+        blockID = OakNativeMemoryAllocator.INVALID_BLOCK_ID;
+        offset = -1;
+        length = -1;
+        version = EntrySet.INVALID_VERSION;
+        buffer = null;
+    }
+
+    boolean isValid() {
```

Comment:
It is very confusing, this "isValid()" relatively to (for example) value state "isValid()". This method just means whether the Slice is initiated or not. Neither it says whether Slice's buffer is initiated or not, nor whether the value/key (in the Slice) is indeed "valid". I mean value can be deleted/not-yet-inserted etc. which is also kind of "not valid". This confusion is prone to errors.

Reply:
Suggestion: slice will remain valid/invalid, but keys/values will be active/inactive.

Reply:
I renamed it to `isAllocated()`.